### PR TITLE
`General`: Fix validation for entity deletion confirmation

### PR DESCRIPTION
**Pull request by zhngharry**
#### General

- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR implements a fix to the bug, that if an entity name is a regex, the confirmation dialog's confirm button is bugged. (Linked Issue: https://github.com/ls1intum/Artemis/issues/8313)

### Description

I fixed the bug by removing the Angular Validators.pattern method and replacing it with a custom simple string comparator function.

### Steps for Testing

Prerequisites:
- 1 Admin/Editor/Instructor

1. Log in to Artemis
2. Navigate to Course Administration
3. Start creating any entity with a regex such as ".*" or "a+" as a name
4. Try to delete the entity and examine whether the delete confirmation button is only activated exactly if the input field string equals the entity name

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Comment by zhngharry** [comment]
This is my first time contributing to Artemis, and in the Contribution Guidelines, it specified that I had to test everything on the test server as well as write own Jest Tests. I had trouble setting up the connection to the test servers, and will have to read up on how to write Jest Tests, and will do so soon.

**Comment by zhngharry** [comment]
Also I believe we should discuss the specific patterns that the titles are validated against; for now I have set everything to the same regex, but maybe the optimal case would be different patterns. Could anyone of the official Artemis Team comment on this?

File: src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise-update.component.html [hunk 1]
```diff
@@ -12,7 +12,14 @@ <h2 id="jhi-text-exercise-heading-import" jhiTranslate="artemisApp.textExercise.
     <jhi-form-status-bar [formStatusSections]="formSectionStatus" />
     <div>
         <h3 jhiTranslate="artemisApp.exercise.sections.general" id="artemisApp.exercise.sections.general"></h3>
-        <jhi-exercise-title-channel-name [exercise]="textExercise" [course]="textExercise.course" [minTitleLength]="3" [isImport]="isImport" [isExamMode]="isExamMode" />
+        <jhi-exercise-title-channel-name
+            [titlePattern]="'^[a-zA-Z0-9- ]+$'"
```
[the remark this answers is not in the document]
Comment (zhngharry) [reply src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise-update.component.html:16]: I am using the same regex pattern everywhere else, I dont understand what could be improved here

**Review by julian-christl — CHANGES_REQUESTED** [review]
Welcome to Artemis. I don't want to discourage you, but I agree with @rstief (https://github.com/ls1intum/Artemis/issues/8313#issuecomment-2088570184): this approach doesn't fix the mentioned issue but makes it obsolete. However, this way, it unnecessarily restricts the freedom of choice an instructor has when naming their course.

The **delete** validation needs an update, not the **create/update** validation.

**Comment by zhngharry** [comment]
> Welcome to Artemis. I don't want to discourage you, but I agree with @rstief ([#8313 (comment)](https://github.com/ls1intum/Artemis/issues/8313#issuecomment-2088570184)): this approach doesn't fix the mentioned issue but makes it obsolete. However, this way, it unnecessarily restricts the freedom of choice an instructor has when naming their course.
> 
> The **delete** validation needs an update, not the **create/update** validation.

Sounds valid, I will revert my changes and implement a fix for the bug itself. Sorry for the troubles ^^

**Comment by julian-christl** [comment]
> Sounds valid, I will revert my changes and implement a fix for the bug itself. Sorry for the troubles ^^

Not at all, I appreciate that you're so relaxed about it and still want to contribute. Hearing "your whole implementation is wrong" is not the news one wants to hear, usually ^^

**Comment by zhngharry** [comment]
> Welcome to Artemis. I don't want to discourage you, but I agree with @rstief ([#8313 (comment)](https://github.com/ls1intum/Artemis/issues/8313#issuecomment-2088570184)): this approach doesn't fix the mentioned issue but makes it obsolete. However, this way, it unnecessarily restricts the freedom of choice an instructor has when naming their course.
> 
> The **delete** validation needs an update, not the **create/update** validation.

I found the spot and fixed it, hopefully this resolves the issue now :)

**Comment by julian-christl** [comment]
Thank you! Could you please update your PR description to reflect the changes? Also include specific examples of what didn't work and should work now in the testing steps.

**Review by rstief — APPROVED** [review]
Code LGTM 👍

**Review by julian-christl — COMMENTED** [review]
The code looks good, thank you. However, the client tests fail on branch coverage. Can you please ensure that the tests cover both cases of this  validator

**Comment by zhngharry** [comment]
> The code looks good, thank you. However, the client tests fail on branch coverage. Can you please ensure that the tests cover both cases of this validator

Did I understand correct that I need to add client tests for this "new" validation logic? Ive looked for the tests and it seems like both cases of the validator are already covered:

src/test/javascript/spec/component/shared/delete-dialog.component.spec.ts:
`it('Form properly checked before submission', async () => {
        // Form can't be submitted if there is deleteConfirmationText and user didn't input the entity title
        comp.entityTitle = 'title';
        comp.deleteConfirmationText = 'artemisApp.exercise.delete.typeNameToConfirm';
        comp.confirmEntityName = '';
        comp.dialogError = new Observable<string>();
        comp.buttonType = ButtonType.ERROR;
        fixture.detectChanges();
        await fixture.whenStable();
        expect(comp.deleteForm.invalid).toBeTrue();
        // TODO: why do changes not propagate to the UI?
        // let submitButton = debugElement.query(By.css('.btn.btn-danger'));
        // expect(submitButton.nativeElement.disabled).toBeTrue();

        // User entered some title
        comp.confirmEntityName = 'some title';
        fixture.detectChanges();
        await fixture.whenStable();
        expect(comp.deleteForm.invalid).toBeTrue();
        // TODO: why do changes not propagate to the UI?
        // submitButton = debugElement.query(By.css('.btn.btn-danger'));
        // expect(submitButton.nativeElement.disabled).toBeTrue();

        // User entered correct tile
        comp.confirmEntityName = 'title';
        fixture.detectChanges();
        await fixture.whenStable();
        expect(comp.deleteForm.invalid).toBeFalse();
        // TODO: why do changes not propagate to the UI?
        // submitButton = debugElement.query(By.css('.btn.btn-danger'));
        // expect(submitButton.nativeElement.disabled).toBeFalse();
    });`
    
    Did i understand something wrong, or why are the checks failing then?

**Comment by julian-christl** [comment]
It's alright. I assumed that it was your PR, but the coverage is failing on develop too.

**Review by julian-christl — APPROVED** [review]
Code looks good

**Comment by julian-christl** [comment]
@zhngharry remember to contact us regarding access to our slack space if you intend to continue contributing, which we would highly appreciate.

https://github.com/ls1intum/Artemis/issues/8313#issuecomment-2088679770

File: src/main/webapp/app/shared/confirm-entity-name/confirm-entity-name.component.ts [hunk 2]
```diff
@@ -89,4 +89,11 @@ export class ConfirmEntityNameComponent implements OnInit, OnDestroy, ControlVal
     registerOnValidatorChange(fn: () => void) {
         this.onValidatorChange = fn;
     }
+
+    private compareWithEntityName(control: FormControl): ValidationErrors | null {
```
Comment (krusche) [inline src/main/webapp/app/shared/confirm-entity-name/confirm-entity-name.component.ts:93]: we prefer `undefined` over `null` in the client code. Could you change or is null necessary due to the Validators API?

Comment (julian-christl) [reply src/main/webapp/app/shared/confirm-entity-name/confirm-entity-name.component.ts:93]: `null` is here indeed the specified return type:
https://angular.io/guide/form-validation#validator-functions

```
constructor(controls: TControl, validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null);
    controls: ɵTypedOrUntyped<TControl, TControl, {
        [key: string]: AbstractControl<any>;
    }>;
```

**Pull request merged**